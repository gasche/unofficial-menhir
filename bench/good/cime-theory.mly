/***************************************************************************

parser for (equational) theory

CiME Project - Démons research team - LRI - Université Paris XI

$Id: theory_parser.mly,v 1.4 2003/09/12 12:29:52 contejea Exp $

***************************************************************************/

%{
  
  open User_signatures
  open Term_algebra

%}

%token <User_signatures.symbol_id> IDENT
%token <int> INT
%token KW_ACU KW_ACI KW_AG KW_ACUN KW_BR
%token COMMA SEMICOLON OPENPAR CLOSEPAR 
%token EOF

%start theory
%type <User_signatures.symbol_id Theory.elem_theory list> theory

%%

theory:
    EOF                    
      { [] }
  | decl 
      { 
        [$1]
      }
  | decl SEMICOLON theory  
      { 
	$1 :: $3 
      }
;

decl:
  acu { $1 }
| aci { $1 }
| ag { $1 }
| acun { $1 }
| br { $1 }
;

acu:
KW_ACU OPENPAR IDENT COMMA IDENT CLOSEPAR 
  { 
    Theory.ACU($3,$5)
  }
;

aci:
KW_ACI OPENPAR IDENT CLOSEPAR 
  { 
    Theory.ACI($3)
  }
;

ag:
KW_AG OPENPAR IDENT COMMA IDENT COMMA IDENT CLOSEPAR 
  { 
    Theory.AG($3,$5,$7)
  }
;

acun:
KW_ACUN OPENPAR IDENT COMMA IDENT COMMA INT CLOSEPAR 
  { 
    Theory.ACUN($3,$5,$7)
  }
;

br:
KW_BR OPENPAR IDENT COMMA IDENT COMMA IDENT COMMA IDENT CLOSEPAR 
  { 
    Theory.BR($3,$5,$7,$9)
  }
;
