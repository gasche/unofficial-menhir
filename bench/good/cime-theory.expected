%{
  
  open User_signatures
  open Term_algebra

%}
%start theory
%token SEMICOLON
%token OPENPAR
%token KW_BR
%token KW_AG
%token KW_ACUN
%token KW_ACU
%token KW_ACI
%token <int> INT
%token <User_signatures.symbol_id> IDENT
%token EOF
%token COMMA
%token CLOSEPAR

%type <User_signatures.symbol_id Theory.elem_theory list> theory
%%

theory:
| EOF
    { [] }
| decl
    { 
        [$1]
      }
| decl SEMICOLON theory
    { 
	$1 :: $3 
      }

decl:
| acu
    { $1 }
| aci
    { $1 }
| ag
    { $1 }
| acun
    { $1 }
| br
    { $1 }

acu:
| KW_ACU OPENPAR IDENT COMMA IDENT CLOSEPAR
    { 
    Theory.ACU($3,$5)
  }

aci:
| KW_ACI OPENPAR IDENT CLOSEPAR
    { 
    Theory.ACI($3)
  }

ag:
| KW_AG OPENPAR IDENT COMMA IDENT COMMA IDENT CLOSEPAR
    { 
    Theory.AG($3,$5,$7)
  }

acun:
| KW_ACUN OPENPAR IDENT COMMA IDENT COMMA INT CLOSEPAR
    { 
    Theory.ACUN($3,$5,$7)
  }

br:
| KW_BR OPENPAR IDENT COMMA IDENT COMMA IDENT COMMA IDENT CLOSEPAR
    { 
    Theory.BR($3,$5,$7,$9)
  }

%%



